### ISSUE TITLE
Compare WSM and WSD annealing

### ISSUE THREAD

**Issue by geoalgo** [OpenEuroLLM/Taskboard#406]
__Goal__
get a rough idea of how WSM works for our checkpoints in downstream performance compared to WSD annealing

__Description__
currently, we have WSD checkpoints at peak learning rate up to 8T tokens
then we have 15 checkpoints from 8 to 8.3T tokens annealed linearly

I propose to check how WSM compares to linear annealing (WSM is a simple approach which which just do weighted average to approximate annealing performance, results reported up to 500BT on 25BT spreadout checkpoints in the [paper](https://arxiv.org/abs/2507.17634))

Of course the comparison will be slightly apple to orange, ideally, we would have checkpoints at peak learning rate from 8 to 8.3T but we just have annealed ones however it can give us an idea whether performance is close and 300BT more should not play a big role given that most metrics are saturated from prelude at this point.

__Deliverable scope__
An evaluation of different weighting proposed in WSM against Prelude checkpoints (both at 8T before annealing and after annealing at 8.3T).

**Comment by geoalgo** [comment]
Steps:

- [x] code to download checkpoints
- [x] code to perform merge
- [ ] push checkpoints on huggingface 
- [ ] evaluate checkpoints on some tasks and compare against baselines

**Comment by geoalgo** [comment]
I ran a quick test on lambdalabs as LUMI was too busy with only 100 samples for each datasets, here are preliminary results:

| Benchmark | No Annealing | Linear Annealing (300BT) | Linear Annealing. (280BT) | Merge Unif. (300BT) | Merge Sqrt (300BT) | Merge Cos. (300BT) |
|---|---|---|---|---|---|---|
| arc_challenge_mt | 43.3 | 44.3 | 44.3 | 43.9 | 43.9 | 43.9 |
| belebele | 60.0 | 68.0 | 64.2 | 67.6 | 67.0 | 68.2 |
| hellaswag_multilingual | 48.4 | 51.2 | 50.9 | 50.6 | 50.3 | 50.7 |
| xcopa | 54.7 | 57.0 | 57.2 | 56.9 | 57.4 | 56.5 |
| include | 43.4 | 46.6 | 45.8 | 42.9 | 40.0 | 43.7 |
| **average** | **50.0** | **53.4** | **52.5** | **52.4** | **51.7** | **52.6** |

<img width="1800" height="900" alt="Image" src="https://github.com/user-attachments/assets/700e2d92-5bd5-4dfc-b95c-8a672341765a" />

The WSM outperforms no annealing and the best approach (Merge Cos. (300BT)) recovers 77% of the gain.
Repeating the description above, the annealed is trained on 300BT more and different sources so it is not apple-to-apple (we dont have checkpoints annealed from 7.7 to 8.0T) but results seems encouraging as we do see general improvement and ballpark of improvement seems correct (the paper reported improvement of merging vs annealing and also found that sqrt was the best whereas it is the worst in our case).

Sounds like a decent way to approximate annealing without spending too much budget / operational effort in our case (in particular for 32B model).

| Variant | Average Recovery |
|---|---|
| Lin. Ann. (280BT) | 0.72 |
| Merge Unif. (300BT) | 0.70 |
| Merge Sqrt (300BT) | 0.51 |
| Merge Cos. (300BT) | 0.77 |

**Comment by geoalgo** [comment]
I will push the code next, merged checkpoints have been pushed to 
https://huggingface.co/openeurollm/prelude-wsm-merges/tree/main
